solr/solrj/src/test/org/apache/solr/common/cloud/TestCollectionStateWatchers.java

```diff
@@ -0,0 +1,235 @@
+
+  @AfterClass
+  public static void shutdownBackgroundExecutors() {
+    executor.shutdown();
+  }
+
+  @Before
+  public void prepareCluster() throws Exception {
+    int missingServers = CLUSTER_SIZE - cluster.getJettySolrRunners().size();
+    for (int i = 0; i < missingServers; i++) {
+      cluster.startJettySolrRunner();
+    }
+    cluster.waitForAllNodes(30);
+  }
+
+  private static Future<Boolean> waitInBackground(String collection, long timeout, TimeUnit unit,
+                                                  CollectionStatePredicate predicate) {
+    return executor.submit(() -> {
+      try {
+        cluster.getSolrClient().waitForState(collection, timeout, unit, predicate);
+      } catch (InterruptedException | TimeoutException e) {
+        return Boolean.FALSE;
+      }
+      return Boolean.TRUE;
+    });
+  }
+
+
+  @Test
+  public void testSimpleCollectionWatch() throws Exception {
+
+    CloudSolrClient client = cluster.getSolrClient();
+    cluster.createCollection("testcollection", CLUSTER_SIZE, 1, "config", new HashMap<>());
+
+    client.waitForState("testcollection", MAX_WAIT_TIMEOUT, TimeUnit.SECONDS, DocCollection::isFullyActive);
+
+    // shutdown a node and check that we get notified about the change
+    final AtomicInteger nodeCount = new AtomicInteger(0);
+    final CountDownLatch latch = new CountDownLatch(1);
+    client.registerCollectionStateWatcher("testcollection", (liveNodes, collectionState) -> {
+      // we can't just count liveNodes here, because that's updated by a separate watcher,
+      // and it may be the case that we're triggered by a node setting itself to DOWN before
+      // the liveNodes watcher is called
+      for (Slice slice : collectionState) {
+        for (Replica replica : slice) {
+          if (replica.isActive(liveNodes))
+            nodeCount.incrementAndGet();
+        }
+      }
+      latch.countDown();
+    });
+
+    cluster.stopJettySolrRunner(random().nextInt(cluster.getJettySolrRunners().size()));
+    assertTrue("CollectionStateWatcher was never notified of cluster change", latch.await(MAX_WAIT_TIMEOUT, TimeUnit.SECONDS));
+
+    assertThat(nodeCount.intValue(), is(3));
+
+  }
+
+  @Test
+  public void testWaitForStateChecksCurrentState() throws Exception {
+
+    CloudSolrClient client = cluster.getSolrClient();
+    cluster.createCollection("waitforstate", 1, 1, "config", new HashMap<>());
+
+    client.waitForState("waitforstate", MAX_WAIT_TIMEOUT, TimeUnit.SECONDS, DocCollection::isFullyActive);
+
+    // several goes, to check that we're not getting delayed state changes
+    for (int i = 0; i < 10; i++) {
+      try {
+        client.waitForState("waitforstate", 1, TimeUnit.SECONDS, DocCollection::isFullyActive);
+      }
+      catch (TimeoutException e) {
+        fail("waitForState should return immediately if the predicate is already satisfied");
+      }
+    }
+
+  }
+
+  @Test
+  public void testCanWatchForNonexistantCollection() throws Exception {
```

Comment:
If there's already a watch set, then it won't be fired.  If there's no watch set, then it will be fired by StateWatcher#refreshAndWatch().